odoo/addons/base/models/qweb.py

```diff
@@ -655,115 +330,214 @@ def _compile_node(self, el, options):
+                        if tokens[index].type == token.NAME:
+                            keys.append(tokens[index].string)
+                        index += 1
+                    is_for_params.append((brackets, keys))
+                index += 1
+
+        is_function_params = [] # list brackets index
+        is_lambda_params = [] # list of tuple (brackets index, keys)
+        brackets = 0
+        index = 0
+        pos = (0, 0) # to keep indent when use expr on multi line
+        while index < tokens_len:
+            t = tokens[index]
+
+            if t.start[0] != pos[0]:
+                pos = (t.start[0], 0)
+            space = t.start[1] - pos[1]
+            if space:
+                result.append(' ' * space)
+            pos = t.start
+
+            if t.type == tokenize.ENCODING or t.type == token.ENDMARKER or t.type == token.DEDENT:
+                pass
+            elif t.type == token.OP and t.string in ['(', '[', '{']: # tuple, function, lambda, for
+                result.append(t.string)
+                brackets += 1
+                if index > 0 and t.string == '(' and tokens[index - 1].type == token.NAME:
+                    is_function_params.append(brackets)
+            elif t.type == token.OP and t.string in [')', ']', '}']: # tuple, function, lambda, for
+                result.append(t.string)
+                if is_function_params and is_function_params[-1] == brackets:
+                    is_function_params.pop()
+                while is_lambda_params and is_lambda_params[-1][0] >= brackets:
+                    is_lambda_params.pop()
+                brackets -= 1
+            elif t.type == token.OP and t.string == ',' and is_lambda_params and is_lambda_params[-1][0] >= brackets: # close lambda
+                while is_lambda_params and is_lambda_params[-1][0] >= brackets:
+                    is_lambda_params.pop()
+                result.append(t.string)
+            elif t.type == token.OP and t.string == '.': # attributes
+                result.append(t.string)
+                if index + 1 < tokens_len and tokens[index + 1].type == token.NAME:
+                    result.append(tokens[index + 1].string)
+                    index += 1
+            elif t.type == token.NAME and t.string == 'lambda': # lambda => allowed values
+                result.append('lambda ')
+                index += 1
+                keys = []
+                while tokens[index] and (tokens[index].type == token.OP or tokens[index].type == token.NAME) and tokens[index].string != ':':
+                    if tokens[index].type == token.NAME:
+                        key = tokens[index].string
+                        result.append(f'_qweb_temp_{key}')
+                        keys.append(key)
+                    else:
+                        result.append(tokens[index].string)
```

Comment:
Handling of non-trivial lambda parameters doesn't work:
* NAME default values are interpreted as lambda parameters
* using the lone `*` placeholder leads to miscompilation

Reply:
If you have examples I would like to add the tests but all of these cases are currently useless for odoo since the model does not use them. In addition, the best practice is to mix as little complex logic as possible in the templates. This is therefore not supposed to happen.

Reply:
> If you have examples I would like to add the tests

`lambda a, *, b=c: (a, b)` shoudl do the trick.

> but all of these cases are currently useless for odoo since the model does not use them. In addition, the best practice is to mix as little complex logic as possible in the templates. This is therefore not supposed to happen.

1. If we don't want to accept it, it should generate a clear and explicit error message, not a mess.
2. We don't control existing templates, we've no idea what clients or partners do in their templates even if non-trivial lambdas would usually not be accepted in our own templates. The pattern of setting a variable as function / lambda default is also a common micro-optimisation as it changes the default `LOAD_NAME`/`LOAD_DEREF`/`LOAD_GLOBAL` to a much faster `LOAD_LOCAL`, even more so for qweb templates as closing over a template variable would not just be a `LOAD_NAME` as the template scope is a dict, it'd be `dict[varname]` aka `LOAD_NAME dict; LOAD_CONST $varname; BINARY_SUBSCR`.